airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialSyncStateIterator.java

```diff
@@ -0,0 +1,98 @@
+package io.airbyte.integrations.source.mysql.initialsync;
+
+import static io.airbyte.integrations.source.mysql.initialsync.MySqlInitialLoadStateManager.MYSQL_STATUS_VERSION;
+
+import autovalue.shaded.com.google.common.collect.AbstractIterator;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.integrations.source.mysql.internal.models.InternalModels.StateType;
+import io.airbyte.integrations.source.mysql.internal.models.PrimaryKeyLoadStatus;
+import io.airbyte.protocol.models.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.AirbyteMessage.Type;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.util.Iterator;
+import java.util.Objects;
+import javax.annotation.CheckForNull;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MySqlInitialSyncStateIterator extends AbstractIterator<AirbyteMessage> implements Iterator<AirbyteMessage> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MySqlInitialSyncStateIterator.class);
+  public static final Duration SYNC_CHECKPOINT_DURATION = Duration.ofMinutes(15);
+  public static final Integer SYNC_CHECKPOINT_RECORDS = 10_000;
+
+  private final Iterator<AirbyteMessage> messageIterator;
+  private final AirbyteStreamNameNamespacePair pair;
+  private boolean hasEmittedFinalState = false;
+  private String lastPk;
+  private final JsonNode streamStateForIncrementalRun;
+  private final MySqlInitialLoadStateManager stateManager;
+  private long recordCount = 0L;
+  private Instant lastCheckpoint = Instant.now();
+  private final Duration syncCheckpointDuration;
+  private final Long syncCheckpointRecords;
+  private final String pkFieldName;
+
+  public MySqlInitialSyncStateIterator(final Iterator<AirbyteMessage> messageIterator,
+      final AirbyteStreamNameNamespacePair pair,
+      final MySqlInitialLoadStateManager stateManager,
+      final JsonNode streamStateForIncrementalRun,
+      final Duration checkpointDuration,
+      final Long checkpointRecords) {
+    this.messageIterator = messageIterator;
+    this.pair = pair;
+    this.stateManager = stateManager;
+    this.streamStateForIncrementalRun = streamStateForIncrementalRun;
+    this.syncCheckpointDuration = checkpointDuration;
+    this.syncCheckpointRecords = checkpointRecords;
+    this.pkFieldName = stateManager.getPrimaryKeyInfo(pair).pkFieldName();
+  }
+
+  @CheckForNull
+  @Override
+  protected AirbyteMessage computeNext() {
+    if (messageIterator.hasNext()) {
+      if ((recordCount >= syncCheckpointRecords || Duration.between(lastCheckpoint, OffsetDateTime.now()).compareTo(syncCheckpointDuration) > 0)
+          && Objects.nonNull(lastPk)
+          && StringUtils.isNotBlank(lastPk)) {
```

Comment:
a blank primary key is still a valid value. Need this check?

Reply:
yeah that's true. Removing